src/NHibernate/Hql/Ast/ANTLR/Tree/DotNode.cs

```diff
@@ -409,12 +412,18 @@ private void DereferenceEntity(EntityType entityType, bool implicitJoin, string
 			}
 			else
 			{
-				joinIsNeeded = generateJoin || ((Walker.IsInSelect && !Walker.IsInCase) || (Walker.IsInFrom && !Walker.IsComparativeExpressionClause));
+				comparisonWithNullableEntity = (Walker.IsComparativeExpressionClause && entityType.IsNullable);
+				joinIsNeeded = generateJoin || (Walker.IsInSelect && !Walker.IsInCase) || (Walker.IsInFrom && !Walker.IsComparativeExpressionClause)
+				               || comparisonWithNullableEntity;
 			}
 
 			if ( joinIsNeeded ) 
 			{
 				DereferenceEntityJoin( classAlias, entityType, implicitJoin, parent );
+				if (comparisonWithNullableEntity)
+				{
+					_columns = FromElement.GetIdentityColumns();
```

Comment:
I'm not sure it's the best way to achieve required result.
This place detects entity comparisons like `where alias.OneToOne is null`. By default such comparisons are translated without join in to something like `where alias.OneToOneId is null`.
But instead we need join and do comparison with joined table `where oneToOneAlias.Id is null` so here we replace for this node columns with identity columns for joined table.